### ISSUE TITLE
Uneditable paragraphs

### ISSUE THREAD

**Issue by ig3** [fedwiki/wiki-client#266]
Sometimes I am unable to edit a paragraph. When I double-click on the paragraph, nothing obvious happens. In particular, the paragraph is not replaced with a textarea containing its content.

When this happens, it is a persistent state. I can edit other paragraphs on the page but when I come back to the uneditable paragraph, it remains uneditable.

After closing and re-opening the page I am able to edit the paragraph again.

Sometimes several paragraphs on a page will be uneditable.

I don't know how to reliably reporduce the problem. It doesn't happen very often.

**Comment by ig3** [comment]
I have just discovered a way to, fairly reliably, reproduce the problem:

1) double-click on a paragraph to edit it
2) double-click on a preceding or following editable paragraph to edit it, while the textarea of the first paragraph remains open

After this, double-click on the first paragraph, whether or not any other paragraph is being edited at the time, has no obvious effect.

**Comment by WardCunningham** [comment]
Each plugin is responsible for handling events and will typically register interest in double-click as they emit and bind their representation in the dom. This happens when a page is viewed and after each markup edit. It is possible that the registration has been lost. Perhaps a javascript error of some type prevented the emit and bind from completing.

Is it always the same plugin? Does it happen with different browsers? Do you see errors in the javascript console?

**Comment by WardCunningham** [comment]
I tried to reproduce the problem with three small paragraphs on a new page using your method. I don't see the problem with an older release of wiki but do see it with a newer release.

**Comment by WardCunningham** [comment]
This is my test case mentioned above.  My test is to refresh the page, double-click the first paragraph, then double-click the second paragraph. I expect the second double-click to close the first editor on loss of focus and then open the the editor on the second paragraph. In both cases I report the javascript console output while clicking. 

This site works correctly.
http://found.ward.bay.wiki.org/view/test
```
textEditor 828a01ea63f4812f {append: true}
textEditor 947c01384a4e6f6e {append: true}
```

This site fails by the second double-click.
http://ward.asia.wiki.org/view/test
```
textEditor 828a01ea63f4812f {append: true}
textEditor 947c01384a4e6f6e {append: true}
notifIndex 0 about to render (3) [div.item.paragraph.ui-sortable-handle, div.item.paragraph.ui-sortable-handle.textEditing, div.item.paragraph.ui-sortable-handle]
```

The second site is running a newer version of wiki-client that includes the search improvements along with other commits. The first site is running older code because I have had some trouble upgrading it. In both cases `wiki -v` reports identical versions but that is clearly not the case.

**Comment by ig3** [comment]
For reference, my site is wiki version 0.21.1, pretty much a default install, other than my modified version of wiki-security-passportjs that authenticates to Azure AD. All the client code is default.

**Comment by WardCunningham** [comment]
Thanks for reporting this problem.

The work around seems to be closing one editor with ESC or CMD-S before opening another.  This might be our habit which explains that we have not noticed this before.

**Comment by ig3** [comment]
Yes, closing with ESC or CMD-S works for me. I discovered these options today when I looked at wiki-client/lib/editor.coffee. I had been clicking outside the page mostly, but occasionally double clicking the next paragraph I wanted to edit.

**Comment by paul90** [comment]
@ig3 Thanks for reporting this. From an initial look it appears that the second textarea is getting removed by a re-render that happens when the first textarea looses focus. In that re-render a class that is set as part of creating the editor is not removed, which causing subsequent attempt to double click to have no effect.

A short term work around, is to click on `wiki` in the footer twice - which removed and then adds the ability to edit.

**Comment by ig3** [comment]
@paul90, click on wiki in the footer twice does restore ability to edit, in my testing.

But, after doing so i notice that the paragraph dblclick handler is registered twice on each paragraph and executes twice when I double click a paragraph. 

After I double click a paragraph then escape, the paragraph and all that follow have the paragraph dblclick handler registered only once but those that precede the paragraph I double clicked still have it registered twice.

I haven't noticed a problem because of the double registration of the dblclick handler, so perhaps it isn't something that needs to be 'fixed', but this suggests to me some flawed logic somewhere. On the other hand, I barely know the code and perhaps my understanding or expectations are not correct. While your change #267 fixes the problem of paragraphs becoming uneditable for me, I am trying to better understand what was happening before the change.

**Comment by paul90** [comment]
The fix for the uneditable paragraphs has been published as + wiki-client@0.20.3

I suspect that double registration of the dblclick handler was introduced when the ability to turn off editing was introduced in #187 (3 years ago). There is a guard in place, that is even older (6 years ago), so that the editor is only added once.

> While your change #267 fixes the problem of paragraphs becoming uneditable for me, I am trying to better understand what was happening before the change.

What appears to be happening when you double click when a text editor is already open is:
1. [textEditor()](https://github.com/fedwiki/wiki-client/blob/master/lib/editor.coffee#L28) is called to create the text editor for the item that has been double click. Amongst other things with registers a `focusout` event on the textarea used. The last thing that happens in `textEditor()` is that focus is moved to the textarea it created.
2. If there was already an editor open, it looses focus and its `focusout` event is fired. Which causes any changes to be saved, and a cascade of re-rendering of page items to be triggered. In this process the textarea that was created in (1) is replaced, but the `textEditing` class on the page item is not - this causes subsequent double clicks on the item to [return](https://github.com/fedwiki/wiki-client/blob/master/lib/editor.coffee#L118) before adding a textarea.
